src/WebJobs.Script.WebHost/Security/SecretManager.cs

```diff
@@ -138,14 +138,78 @@ public virtual HostSecretsInfo GetHostSecrets()
                         }
                     };
 
-                    string secretsContent = ScriptSecretSerializer.SerializeFunctionSecrets(secrets);
-                    File.WriteAllText(secretFilePath, secretsContent);
+                    PersistFunctionSecrets(secrets, secretsFilePath);
+                }
+
+                // Read all secrets, which will run the keys through the appropriate readers
+                var keys = secrets.Keys.Select(k => ReadSecretValue(k)).ToList();
+
+                if (keys.Any(k => k.PersistenceStateStale))
+                {
+                    UpdateFunctionSecretsPersistenceState(keys, secretsFilePath);
                 }
 
-                return secrets.Keys.ToDictionary(s => s.Name, s => ReadSecretValue(s));
+                return secrets.Keys.ToDictionary(s => s.Name, s => s.Value);
             });
         }
 
+        private HostSecrets GenerateHostSecrets(string secretsFilePath)
+        {
+            // initialize with new secrets and save it
+            var hostSecrets = new HostSecrets
+            {
+                MasterKey = GenerateSecret(DefaultMasterKeyName),
+                FunctionKeys = new List<Key>
+                {
+                    GenerateSecret(DefaultFunctionKeyName)
+                }
+            };
+
+            PersistHostSecrets(hostSecrets, secretsFilePath);
+
+            return hostSecrets;
+        }
+
+        private void UpdateFunctionSecretsPersistenceState(List<Key> keys, string secretsFilePath)
```

Comment:
Is `RefreshFunctionSecrets` clearer? The `SecretManager` doesn't perform or is aware of encryption, so that method doesn't really know if the content will be encrypted, just that it needs to be updated and persisted.